PowerEditor/src/WinControls/Grid/ShortcutMapper.cpp

```diff
@@ -124,123 +122,176 @@ void ShortcutMapper::initBabyGrid() {
 	_babygrid.setHighlightColorProtectNoFocus(RGB(230,194,190));
 }
 
+generic_string ShortcutMapper::getTextFromCombo(HWND hCombo)
+{
+	TCHAR str[128];
+	::SendMessage(hCombo, WM_GETTEXT, 127, reinterpret_cast<LPARAM>(str));
+	generic_string res(str);
+	return stringToLower(res);
+};
+
+bool ShortcutMapper::isFilterValid(Shortcut sc)
+{
+	bool match;
+	generic_string shortcut_name = stringToLower(generic_string(sc.getName()));
+	match = false;
+	if (_shortcutFilter[0] == TEXT('\0')){

```

Comment:
The [TEXT macro](https://msdn.microsoft.com/de-de/library/windows/desktop/dd374074%28v=vs.85%29.aspx) is only defined for strings, not for individual characters. You should get a compiler warning here and the program should crash immediately.

Reply:
@MAPJe71 Yes, I'm sure that the `TEXT` macro is only defined for strings. That's what the documentation says.

By the way, the same macro can be used as follows:

    TEXT(oadLibrary)("user32.dll")

It works, but is obviously not correct.

Reply:
@rillig The documentation does not say _only_ for strings.
I've been using it for years based on (old) MS documentation for both strings and characters (see also comment in `tchar.h` code snippet above) and never had a compiler warning (only used MSVS compilers).
What compiler gives the warning and what is the code?

Reply:
@MAPJe71 There's no compiler warning I know of. I just thought that based on the definition of the macro in the MSDN there should be a type conflict, since the definition there says `void TEXT(LPTSTR string)`. But since it is a macro and not a function, that though of mine was wrong.

Reply:
@MAPJe71 Just to be pedantic, the MSDN documentation clearly says _only for strings_, repeatedly:

* Identifies a **string** as Unicode
* Otherwise, the macro identifies a **string**
* void TEXT(**LPTSTR** string)
* Pointer to the **string**

Just because it _apparently works_ for single characters too doesn't mean it's guaranteed; see my `oadLibrary` example, which also works but isn't guaranteed.

The documentation does not explicitly need to mention the word _only_ because otherwise this word would have to appear too often and reading the text would get boring.

Reply:
Depends on what (part of the) documentation one is reading ([String and Character Literals](https://docs.microsoft.com/en-us/cpp/cpp/string-and-character-literals-cpp)).

Reply:
@MAPJe71 The documentation you cite only talks about the `L` modifier. It doesn't talk about the `TEXT` macro. This distinction was my point of argument the whole time. I know that you can write `L'a'` to get a wide character but that in itself and the knowledge that the `TEXT` macro is currently implemented such that it produces this sequence doesn't make it valid code. The `TEXT` macro might also be implemented in a way that produces a compiler-specific string literal that is only valid for strings but not for characters. And then the program blows up. Hopefully during compilation and not during runtime.

Reply:
> The TEXT macro might also be implemented in a way that produces a compiler-specific string literal that is only valid for strings but not for characters.

True but any macro or function _might_ be implemented differently in the future.
The documentation _might_ not be up to date or incomplete. Although the same can be said about the source code of the library.

Do a search for `TEXT('` in the Microsoft header files and be supprised how often they use the `TEXT` macro for character literals, apparently contradicting their own documentation (e.g. `sddl.h`, `strsafe.h`).

Reply:
@MAPJe71 Ok, you won. That last paragraph is very convincing. :)

Reply:
big thread ! 
So my code is good, isn't it ?

Reply:
@cmeriaux No, your code is too complicated. Since you are not really interested in the first character of the string, you shouldn't ask for it. The only thing you want to know is whether the string is empty. So just ask `_shortcutFilter.empty()`.